File: CP/Hiring_contests/codechef_intern_oct20/sadsale.cpp
#include<bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef unsigned long long ull;
typedef long double ld;
#define X first
#define Y second
#define endl "\n"
const ll mod = 1e9+7;
const int mxn = 2e5+2;

vector<pair<int,ll>> gr[mxn];
int anc[mxn][20];
int level[mxn];
ll psum[mxn];

int n,q,r;

void dfs(int s,int par){
    for (int i=1; i<20; i++)
        anc[s][i] = anc[anc[s][i-1]][i-1];
    for(auto ent:gr[s]){
        int i = ent.X;
        ll w = ent.Y;
        if (i==par)continue;
        psum[i] = psum[s]+w;
        level[i] = level[s]+1;
        anc[i][0] = s;
        dfs(i,s);
    }
}

int lca(int u, int v){
    if (level[u]>level[v]) swap(u,v);
    for (int i=19; i>=0; i--){
        if (level[v] - (1<<i) >= level[u]) v = anc[v][i];
    }
    if (u==v)return u;
    for (int i=19; i>=0; i--){
        if (anc[u][i]!=anc[v][i]){
            u = anc[u][i];
            v = anc[v][i];
        }
    }
    return anc[u][0];
}

ll pathsum(int u, int v){
    int lc = lca(u,v);
    return psum[u]+psum[v]-2*psum[lc];
}


void solve(){
    cin>>n>>q>>r;
    for( int i =0 ; i<=n; ++i){
        for( int j = 0; j<=20; ++j)
            anc[i][j] = -1;
        gr[i].clear();
        psum[i] = 0;
    }
    int u,v;
    ll w;
    for (int i =0 ; i<n-1; ++i){
        cin>>u>>v>>w;
        gr[u].push_back({v,w});
        gr[v].push_back({u,w});
    }
    level[r] = 0;
    dfs(r,-1);
    for(int i = 0; i<q; ++i){
        cin>>u>>v;
        cout<<pathsum(u,v)<<endl;
    }
}

int main(){
    // freopen("in.txt", "r", stdin);
    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    int t = 1;
    cin>>t;
    while(t-->0) solve();
    return 0;
}

File: CP/Hiring_contests/codechef_intern_oct20/sadsale.py
import os, sys, bisect
from collections import defaultdict, Counter, deque;
from functools import lru_cache   #use @lru_cache(None)
if os.path.exists('in.txt'): sys.stdin=open('in.txt','r')
if os.path.exists('out.txt'): sys.stdout=open('out.txt', 'w')
#
input = lambda: sys.stdin.readline().strip()
imap = lambda: map(int,input().split()); ilist = lambda: list(imap())
#------------------------------------------------------------------
sys.setrecursionlimit(10**6)


for _ in range(int(input())):
    n,q,r = imap()
    gr = defaultdict(list)
    for i in range(n-1):
        u,v,w = imap()
        gr[u].append([v,w])
        gr[v].append([u,w])
    anc = [[-1]*(20) for i in range(n+1)]
    level = [0]*(n+1)
    psum = [0]*(n+1)
    def dfs(s,par = -1):
        for i in range(19,0,-1):
            anc[s][i] = anc[anc[s][i-1]][i-1]
        for i,w in gr[s]:
            if i==par: continue
            level[i] = level[s]+1
            anc[i][0] = s
            psum[i] = psum[s]+w
            dfs(i,s)
    dfs(r)
    def lca(u,v):
        if level[u]>level[v]: u,v = v,u
        for i in range(19,-1,-1):
            if level[v]-(1<<i)>=level[u]:
                v = anc[v][i]
        if u==v: return v
        for i in range(19, -1,-1):
            if anc[u][i]!=anc[v][i]:
                u = anc[u][i]
                v = anc[v][i]
        return anc[u][0]
    def pathsum(u,v):
        lc = lca(u,v)
        return psum[u]+psum[v]-2*psum[lc]

    res = []
    for i in range(q):
        u,v = imap()
        res.append(str(pathsum(u,v)))
    print("\n".join(res))

